File: 322.零钱兑换.cpp
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */
#include "algorithm"
#include "iostream"
#include "map"
#include "queue"
#include "vector"

using namespace std;
// @lc code=start
class Solution {
  private:
    map<int, int> m;

  public:
    int coinChange(vector<int> &coins, int amount) { return dp(amount, coins); }

    int dp(int n, vector<int> &coins) {
        if (m.count(n)) {
            return m[n];
        }
        if (n < 0)
            return -1;
        if (n == 0)
            return 0;
        int res = INT32_MAX;
        for (auto &i : coins) {
            int sub = dp(n - i, coins);
            if (sub == -1) {
                continue;
            } else {
                res = min(res, 1 + sub);
            }
        }
        res = res != INT32_MAX ? res : -1;
        m[n] = res;
        return res;
    }
};
// @lc code=end

File: 322.零钱兑换.go
/*
 * @lc app=leetcode.cn id=322 lang=golang
 *
 * [322] 零钱兑换
 */

// @lc code=start

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func coinChange(coins []int, amount int) int {
	var dp = make([]int, amount + 1)
	for i := 0; i < amount + 1; i++ {
		dp[i] = amount + 1
	}
	dp[0] = 0
	for i := 1; i < amount + 1; i++ {
		for _, c := range coins {
			if i - c < 0 {
				continue
			}
			dp[i] = min(dp[i], dp[i - c] + 1)
		}
	}
	if dp[amount] == amount + 1 {
		return -1
	}
	return dp[amount]
}
// @lc code=end

